src/utils/use-optimistic-search-params-state.ts

```diff
@@ -0,0 +1,31 @@
+import { useNavigation } from '@remix-run/react';
+import { useCallback, useEffect, useState } from 'react';
+
+/**
+ * Allows updating the UI optimistically while Remix navigates to the URL with updated parameters.
+ */
+export function useOptimisticSearchParamsState<S>(
+    stateFromSearchParams: S,
+    setStateToSearchParams: (state: S) => void,
```

Comment:
Same ambiguity as above. I'm giving it state, so I guess this callback should update search params? But if the hook is only for updating UI optimistically, why is it changing search params?

I'm quite confused about what this hook is responsible for.

Reply:
> setStateToSearchParams - same ambiguity as above, I'm giving this function state, so I guess it should update search params? But if the hook is only for updating UI optimistically, why is it changing search params?

Not sure I get it.

The hook is responsible for providing react state that we can update immediately and sync it with the state in URL search params.

Reply:
The part "keep it in sync it with the state in URL search params" is not clear to me.

I'm looking at the params, at the return value `return [state, handleChange]`, at the function description, and things don't connect in my head.